Auton recorder

kirstenalice opened the pull request:
beginnings of the auton recorder
not finished, string to frame and frame to string conversion is functional
reading and writing to file is not done and not tested

Commit:
added AutonRecorder files

Commit:
basic structure for auton recorder

Commit:
basic structure for auton recorder

Commit:
progress check on auton recorder

mostly untested, substring and formatString (method unfinished) work

Commit:
stringToFrame is functional!!

Commit:
renamed some functions

Commit:
frame to string and string to frame conversion works!!

vex750evolution commented on 2016-2017_Starstruck/Module Test Code/FileIO_Test/include/AutonRecorder.h:
```diff
@@ -0,0 +1,87 @@
+
+  typedef struct Frame {
+      int analog_main[4]; //ch1, ch2, ch3, ch4
+      bool digital_main[12]; //5U, 5D, 6U, 6D, 7U, 7D, 7L, 7R, 8U, 8D, 8L, 8R
```
Might want to add macros for these to make it easier so we can do something like frame.digital_main[J_5U] to get values.

vex750evolution commented on 2016-2017_Starstruck/Module Test Code/FileIO_Test/include/AutonRecorder.h:
```diff
@@ -0,0 +1,87 @@
+
+
+  static const Frame NULLFRAME = {
+    {-1,-1,-1,-1},
```
when defining structs, you want to try to use the more explicit syntax when possible. i.e.:

```
static const Frame NULLFRAME = {
    .analog_main = {-1, -1, -1, -1},
...
```

That way you can easily tell what's being assigned to what and you don't have to memorize the struct order

vex750evolution commented on 2016-2017_Starstruck/Module Test Code/FileIO_Test/src/AutonRecorder.c:
```diff
@@ -0,0 +1,193 @@
+  char * frameToString(Frame frame){
+    char * string = malloc(50);
+    sprintf(string,"%d%d%d%d%d%d%d%d%d%d%d%d%d%d%d%d%d%d%d%d%d%d%d%d%d%d%d%d%d%d%d%d",
+      frame.analog_main[0]+127,
```
see comment in header file about macros, these array indeces are not self-explanatory at all

vex750evolution commented on 2016-2017_Starstruck/Module Test Code/FileIO_Test/src/AutonRecorder.c:
```diff
@@ -0,0 +1,193 @@
+  }
+
+  char * frameToString(Frame frame){
+    char * string = malloc(50);
```
don't hardcode number of bytes. Use `malloc(sizeof(char)*strLength)`

vex750evolution commented on 2016-2017_Starstruck/Module Test Code/FileIO_Test/src/AutonRecorder.c:
```diff
@@ -0,0 +1,193 @@
+
+    int x;
+    for(x=0;x<4;x++){
+      frame.analog_main[x] = atoi(substring(analog_main,(x*3)+1,3))-127;
```
might want to add comments to explain this math cuz we're not gonna remember what it means later

vex750evolution commented on 2016-2017_Starstruck/Module Test Code/FileIO_Test/src/AutonRecorder.c:
```diff
@@ -0,0 +1,193 @@
+       exit(1);
+    }
+
+    for(subPos = 0 ; subPos < length ; subPos++){
```
not a huge fan of this implementation, I think the strncpy version on #electrical will be cleaner and faster (worth it to benchmark speeds or nah?)

kirstenalice replied:
i tried using it but it doesnt work lol

matticoli replied:
I threw the code I sent together using nano and gcc, didn't test on the cortex. If this works we can take a look at it later, I'll open an issue for it

matticoli replied:
See #30

kirstenalice replied:
this is what u gave me before it doesn't work we can look at it tomorrow

vex750evolution commented on 2016-2017_Starstruck/Module Test Code/FileIO_Test/src/opcontrol.c:
```diff
@@ -38,8 +38,7 @@
  * ======== */
 
 void operatorControl() {
```
I'm assuming all of this is just test code?

vex750evolution commented on 2016-2017_Starstruck/Module Test Code/FileIO_Test/include/AutonRecorder.h:
```diff
@@ -0,0 +1,87 @@
+  };
+
+  void startRecording(int slot);
+  /*
```
docs should go above function, and use @param annotations to explain parameters (see docs format in API.h or LCD.h)  The reason for this is that the IDE will pick up on it and show the explanations in the autocomplete popups. We can also use it later to auto-generate an API reference for the modules

kirstenalice replied:
idk what ur talking about

vex750evolution commented on 2016-2017_Starstruck/Module Test Code/FileIO_Test/include/main.h:
```diff
@@ -44,6 +44,9 @@
 
 #include <API.h>
+#include <AutonRecorder.h>
+#include <stdlib.h>
```
I think stdlib is included in API.h, but not sure. string.h should be included in AutonRecorder.h to make sure dependencies are managed properly (if this is being imported into a new project, you don't want the user to have to worry about adding system imports that your module is dependent on)

vex750evolution commented on 2016-2017_Starstruck/Module Test Code/FileIO_Test/src/AutonRecorder.c:
```diff
@@ -0,0 +1,193 @@
+ #include <main.h>
+
+/*
+ * This file abstracts all of the functionality of 750E's Auton Recorder.
```
File descriptions are gonna have to be tweaked but that's a later problem

vex750evolution commented on 2016-2017_Starstruck/Module Test Code/FileIO_Test/src/AutonRecorder.c:
```diff
@@ -0,0 +1,193 @@
+  void addFrame(Frame toAdd){
+      toAdd.previous = root.previous;
+      toAdd.next = &root;
+      root.previous = &toAdd;
```
need to get root.previous before changing so you can set its next to toAdd, you're missing a link

vex750evolution commented on 2016-2017_Starstruck/Module Test Code/FileIO_Test/src/AutonRecorder.c:
```diff
@@ -0,0 +1,193 @@
+  }
+
+  void recordingLoop(){
+    if(root.analog_main[0] == -1){ //checks if it's equal to NULLFRAME
```
any reason you did this here and not in startRecording? Might actually be better this way but I was wondering if you had a specific reason

kirstenalice replied:
did what? checking if the root is null or assigning the root a value?

matticoli replied:
assigning it a value. This actually works better so you don't have to pass a reference to the root out of the init function (see discussion on #28 )

kirstenalice replied:
idk what youre talking about

vex750evolution reviewed (commented):
Looks great so far, just commented in a few notes and some convention/semantics fixes. 

Reviewed by @AsciiPorpoise 
(Forgot I was on the team account when I started writing this review and I don't want to redo it)

Commit:
changed some stuff from PR, deleted commented test code

Commit:
deleted comments

Commit:
reads and writes values to file successfully!!

Commit:
changed testing strings

Commit:
trying to read and write values from file in stopRecording

changed value of NULLFRAME
i think there are problems with the pointers to the nexts and prevs in
the Frames whats a pointer idk

Commit:
fixed file writing

Commit:
loop works, but it wont read joystick values

Commit:
now reads joystick values, not finished

Commit:
creates frames based on joystick values, issue with nexts and previous's for frame linked list, not writing every frame to file

Commit:
made functions take in pointers to Frames, not passing by value anymore but not fully functional

Commit:
root initialization not working

Commit:
fixed task management in stoprecording

kirstenalice committed:
Merge branch 'auton-recorder' into auton-task-fix

kirstenalice committed:
Merge pull request #32 from SBHSRobotics/auton-task-fix

fixed task management in stoprecording

Commit:
fixed compile errors except for non-scalar

Commit:
fixed more compile errors, functions that return Frame now return Frame*

Commit:
pointer nonsense in printAllFrames

Commit:
printAllFrames works!!!!

Commit:
stopRecording works!!!!!!!

srikayy committed:
Header/Skeleton for replayer

srikayy committed:
Merge branch 'auton-recorder' into auton-replayer

Commit:
auton recorder is perfect af

srikayy committed:
Merge branch 'auton-recorder' into auton-replayer

srikayy committed:
input function basics

srikayy committed:
working changes

Commit:
inputGetDigital written and warnings removed

srikayy committed:
replayerLoop done

Commit:
Merge remote-tracking branch 'origin/auton-replayer' into auton-replayer

# Conflicts:
#	2016-2017_Starstruck/Module Test Code/FileIO_Test/src/AutonReplayer.c

srikayy committed:
Replayer ready to test

Commit:
moved recordingLoop from autonRecorder h file to c file

srikayy committed:
opcontrol test code for replayer

srikayy committed:
Merge branch 'auton-replayer' of https://github.com/SBHSRobotics/750E into auton-replayer

Commit:
created saveRecording and changed stopRecording

stopRecording is now saveRecording, stopRecording is a new function

Commit:
added break in opcontrol loop

Commit:
put together recorder and replayer

srikayy committed:
Testing changes

srikayy committed:
Working changes

Commit:
text formatting

matticoli committed:
Merge pull request #33 from SBHSRobotics/auton-replayer

Auton replayer

Commit:
still doesnt work, still weird strings :(

Commit:
big changes

deleted linked list
combined saveRecording and startRecording
file writing doesnt work anymore what the FUCK

Commit:
auton recorder works!!!

Commit:
deleted robot code

matticoli committed:
removed extraneous comma

matticoli committed:
removed addFrame from header

matticoli committed:
fixed replayerloop logic to update currentFrame
